File: src/services/payment.service.ts
import crypto from 'crypto';
import { razorpay } from '../config/razorpay';
import { prisma } from '../config/prisma';
import { AppError } from '../middleware/error';
import { PaymentMethod, PaymentStatus } from '@prisma/client';

export class PaymentService {
  public static async createOrder(data: {
    amount: number;
    currency: string;
    method: PaymentMethod;
    customerEmail: string;
    customerPhone?: string | null;
  }) {
    const { amount, currency, method, customerEmail, customerPhone } = data;

    if (method === PaymentMethod.COD) {
      const orderId = `COD-${crypto.randomBytes(8).toString('hex').toUpperCase()}`;
      
      const payment = await prisma.payment.create({
        data: {
          orderId,
          amount,
          currency,
          method,
          status: PaymentStatus.PENDING,
          customerEmail,
          customerPhone,
        },
      });

      return {
        orderId: payment.orderId,
        amount: Number(payment.amount),
        currency: payment.currency,
        method: payment.method,
        status: payment.status,
      };
    } else {
      // Create Razorpay Order
      try {
        const amountInPaise = Math.round(amount * 100);
        const razorpayOrder = await razorpay.orders.create({
          amount: amountInPaise,
          currency: currency,
          receipt: `rcpt_${crypto.randomBytes(4).toString('hex')}`,
        });

        const payment = await prisma.payment.create({
          data: {
            orderId: razorpayOrder.id,
            amount: amount,
            currency: currency,
            method,
            status: PaymentStatus.PENDING,
            customerEmail,
            customerPhone,
          },
        });

        return {
          orderId: payment.orderId,
          amount: amountInPaise, // Returned in paise for Checkout SDK frontend use
          currency: payment.currency,
          keyId: process.env.RAZORPAY_KEY_ID,
          method: payment.method,
          status: payment.status,
        };
      } catch (error: any) {
        const errorMsg = error.error?.description || error.message || JSON.stringify(error);
        throw new AppError(`Razorpay Order Creation Failed: ${errorMsg}`, 500);
      }
    }
  }

  public static async verifySignature(data: {
    razorpay_order_id: string;
    razorpay_payment_id: string;
    razorpay_signature: string;
  }) {
    const { razorpay_order_id, razorpay_payment_id, razorpay_signature } = data;

    // Verify signature logic: HMAC-SHA256(order_id + "|" + payment_id, secret)
    const secret = process.env.RAZORPAY_KEY_SECRET!;
    const text = `${razorpay_order_id}|${razorpay_payment_id}`;
    const generatedSignature = crypto
      .createHmac('sha256', secret)
      .update(text)
      .digest('hex');

    if (generatedSignature !== razorpay_signature) {
      try {
        await prisma.payment.update({
          where: { orderId: razorpay_order_id },
          data: { status: PaymentStatus.FAILED },
        });
      } catch (e) {
        // Payment entry might not exist
      }
      throw new AppError('Payment verification failed: Signature mismatch', 400);
    }

    try {
      const updatedPayment = await prisma.payment.update({
        where: { orderId: razorpay_order_id },
        data: {
          paymentId: razorpay_payment_id,
          signature: razorpay_signature,
          status: PaymentStatus.CAPTURED,
        },
      });

      // Sync Order status to PAID if linked
      if (updatedPayment.orderRefId) {
        await prisma.order.update({
          where: { id: updatedPayment.orderRefId },
          data: { status: 'PAID' },
        });
      }

      return updatedPayment;
    } catch (error: any) {
      throw new AppError('Payment not found or database update failed', 404);
    }
  }

  public static async handleWebhook(rawBody: string, signature: string) {
    const webhookSecret = process.env.RAZORPAY_WEBHOOK_SECRET;
    if (!webhookSecret) {
      throw new AppError('Webhook secret not configured', 500);
    }

    // Verify webhook signature
    const expectedSignature = crypto
      .createHmac('sha256', webhookSecret)
      .update(rawBody)
      .digest('hex');

    if (expectedSignature !== signature) {
      throw new AppError('Invalid webhook signature', 400);
    }

    let event;
    try {
      event = JSON.parse(rawBody);
    } catch (error) {
      throw new AppError('Invalid Webhook JSON payload', 400);
    }

    if (event.event === 'payment.captured') {
      const paymentPayload = event.payload.payment.entity;
      const orderId = paymentPayload.order_id;
      const paymentId = paymentPayload.id;

      if (orderId) {
        try {
          const payment = await prisma.payment.findUnique({
            where: { orderId },
          });

          if (payment) {
            await prisma.payment.update({
              where: { orderId },
              data: {
                status: PaymentStatus.CAPTURED,
                paymentId: paymentId,
              },
            });

            if (payment.orderRefId) {
              await prisma.order.update({
                where: { id: payment.orderRefId },
                data: { status: 'PAID' },
              });
            }
          }
        } catch (error) {
          console.error('[Webhook Error] Failed to update payment/order status:', error);
        }
      }
    }

    return { received: true };
  }
}

File: backend/src/main/java/com/razorpay/integration/service/PaymentService.java
package com.razorpay.integration.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.razorpay.Order;
import com.razorpay.Payment;
import com.razorpay.RazorpayClient;
import com.razorpay.integration.dto.CreatePaymentRequest;
import com.razorpay.integration.dto.PaymentResponseDto;
import com.razorpay.integration.dto.VerifyPaymentRequest;
import com.razorpay.integration.exception.AppException;
import com.razorpay.integration.model.*;
import com.razorpay.integration.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.security.SecureRandom;
import java.util.HexFormat;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final CartItemRepository cartItemRepository;
    private final WhatsAppService whatsAppService;
    private final RazorpayClient razorpayClient;
    private final ObjectMapper objectMapper;

    @Value("${razorpay.key-id}")
    private String keyId;

    @Value("${razorpay.key-secret}")
    private String keySecret;

    @Value("${razorpay.webhook-secret:yourWebhookSecret}")
    private String webhookSecret;

    private static final SecureRandom RANDOM = new SecureRandom();

    public PaymentMethod mapRazorpayMethodToEnum(String razorpayMethod) {
        if (razorpayMethod == null) return PaymentMethod.CARD;
        String upper = razorpayMethod.toUpperCase();
        return switch (upper) {
            case "UPI" -> PaymentMethod.UPI;
            case "CARD" -> PaymentMethod.CARD;
            case "NETBANKING", "BANK" -> PaymentMethod.NETBANKING;
            case "WALLET" -> PaymentMethod.WALLET;
            default -> PaymentMethod.CARD;
        };
    }

    @Transactional
    public PaymentResponseDto createPayment(CreatePaymentRequest request) {
        BigDecimal amount = request.getAmount();
        String currency = request.getCurrency() != null ? request.getCurrency() : "INR";
        PaymentMethod method = request.getMethod();

        if (method == PaymentMethod.COD) {
            byte[] bytes = new byte[8];
            RANDOM.nextBytes(bytes);
            String orderId = "COD-" + HexFormat.of().formatHex(bytes).toUpperCase();

            com.razorpay.integration.model.Payment payment = com.razorpay.integration.model.Payment.builder()
                    .orderId(orderId)
                    .amount(amount)
                    .currency(currency)
                    .method(method)
                    .status(PaymentStatus.PENDING)
                    .customerEmail(request.getCustomerEmail())
                    .customerCountryCode(request.getCustomerCountryCode() != null ? request.getCustomerCountryCode() : "+91")
                    .customerPhone(request.getCustomerPhone())
                    .build();

            payment = paymentRepository.save(payment);

            return PaymentResponseDto.builder()
                    .orderId(payment.getOrderId())
                    .amount(payment.getAmount())
                    .currency(payment.getCurrency())
                    .method(payment.getMethod())
                    .status(payment.getStatus())
                    .build();
        } else {
            try {
                long amountInPaise = amount.multiply(BigDecimal.valueOf(100)).longValue();

                JSONObject orderRequest = new JSONObject();
                orderRequest.put("amount", amountInPaise);
                orderRequest.put("currency", currency);
                orderRequest.put("receipt", "rcpt_" + UUID.randomUUID().toString().substring(0, 8));

                Order razorpayOrder = razorpayClient.orders.create(orderRequest);
                String razorpayOrderId = razorpayOrder.get("id");

                com.razorpay.integration.model.Payment payment = com.razorpay.integration.model.Payment.builder()
                        .orderId(razorpayOrderId)
                        .amount(amount)
                        .currency(currency)
                        .method(method)
                        .status(PaymentStatus.PENDING)
                        .customerEmail(request.getCustomerEmail())
                        .customerCountryCode(request.getCustomerCountryCode() != null ? request.getCustomerCountryCode() : "+91")
                        .customerPhone(request.getCustomerPhone())
                        .build();

                payment = paymentRepository.save(payment);

                return PaymentResponseDto.builder()
                        .orderId(payment.getOrderId())
                        .amount(amountInPaise)
                        .currency(payment.getCurrency())
                        .keyId(keyId)
                        .method(payment.getMethod())
                        .status(payment.getStatus())
                        .build();
            } catch (Exception e) {
                log.error("Razorpay Order Creation Failed: {}", e.getMessage(), e);
                throw new AppException("Razorpay Order Creation Failed: " + e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
            }
        }
    }

    @Transactional
    public void completeOnlineOrder(Integer orderId, PaymentMethod finalMethod) {
        com.razorpay.integration.model.Order order = orderRepository.findById(orderId).orElse(null);
        if (order == null || order.getStatus() == OrderStatus.PAID) {
            return;
        }

        // Decrement product inventory
        for (OrderItem item : order.getItems()) {
            Product product = item.getProduct();
            product.setStock(Math.max(0, product.getStock() - item.getQuantity()));
            productRepository.save(product);
        }

        // Clear Cart items
        if (order.getCartId() != null) {
            cartItemRepository.deleteByCartId(order.getCartId());
        }

        // Update Order status
        order.setStatus(OrderStatus.PAID);
        order.setPaymentMethod(finalMethod);
        orderRepository.save(order);

        // Send WhatsApp notification
        String customerName = order.getUser() != null ? order.getUser().getName() : "Customer";
        String toPhone = (order.getCustomerCountryCode() != null ? order.getCustomerCountryCode() : "+91") + order.getCustomerPhone();
        whatsAppService.sendOrderConfirmation(toPhone, customerName, order.getOrderNumber(), order.getTotalAmount(), order.getShippingAddress());
    }

    @Transactional
    public com.razorpay.integration.model.Payment verifySignature(VerifyPaymentRequest request) {
        String razorpayOrderId = request.getRazorpay_order_id();
        String razorpayPaymentId = request.getRazorpay_payment_id();
        String razorpaySignature = request.getRazorpay_signature();

        String payload = razorpayOrderId + "|" + razorpayPaymentId;
        String generatedSignature = calculateHmacSha256(payload, keySecret);

        if (!generatedSignature.equalsIgnoreCase(razorpaySignature)) {
            paymentRepository.findByOrderId(razorpayOrderId).ifPresent(p -> {
                p.setStatus(PaymentStatus.FAILED);
                paymentRepository.save(p);
            });
            throw new AppException("Payment verification failed: Signature mismatch", HttpStatus.BAD_REQUEST);
        }

        PaymentMethod finalMethod = PaymentMethod.CARD;
        try {
            Payment razorpayPayment = razorpayClient.payments.fetch(razorpayPaymentId);
            if (razorpayPayment != null && razorpayPayment.has("method")) {
                finalMethod = mapRazorpayMethodToEnum(razorpayPayment.get("method"));
            }
        } catch (Exception e) {
            log.error("Failed to fetch payment details from Razorpay: {}", e.getMessage());
        }

        com.razorpay.integration.model.Payment payment = paymentRepository.findByOrderId(razorpayOrderId)
                .orElseThrow(() -> new AppException("Payment record not found for orderId: " + razorpayOrderId, HttpStatus.NOT_FOUND));

        payment.setPaymentId(razorpayPaymentId);
        payment.setSignature(razorpaySignature);
        payment.setStatus(PaymentStatus.CAPTURED);
        payment.setMethod(finalMethod);
        payment = paymentRepository.save(payment);

        if (payment.getOrderRef() != null) {
            completeOnlineOrder(payment.getOrderRef().getId(), finalMethod);
        }

        return payment;
    }

    @Transactional
    public Map<String, Object> handleWebhook(String rawBody, String signature) {
        if (webhookSecret == null || webhookSecret.trim().isEmpty()) {
            throw new AppException("Webhook secret not configured", HttpStatus.INTERNAL_SERVER_ERROR);
        }

        String expectedSignature = calculateHmacSha256(rawBody, webhookSecret);
        if (!expectedSignature.equalsIgnoreCase(signature)) {
            throw new AppException("Invalid webhook signature", HttpStatus.BAD_REQUEST);
        }

        try {
            JsonNode root = objectMapper.readTree(rawBody);
            String eventType = root.path("event").asText();

            if ("payment.captured".equals(eventType)) {
                JsonNode paymentEntity = root.path("payload").path("payment").path("entity");
                String orderId = paymentEntity.path("order_id").asText();
                String paymentId = paymentEntity.path("id").asText();
                String razorpayMethod = paymentEntity.path("method").asText();
                PaymentMethod finalMethod = mapRazorpayMethodToEnum(razorpayMethod);

                if (orderId != null && !orderId.isEmpty()) {
                    paymentRepository.findByOrderId(orderId).ifPresent(payment -> {
                        payment.setStatus(PaymentStatus.CAPTURED);
                        payment.setPaymentId(paymentId);
                        payment.setMethod(finalMethod);
                        paymentRepository.save(payment);

                        if (payment.getOrderRef() != null) {
                            completeOnlineOrder(payment.getOrderRef().getId(), finalMethod);
                        }
                    });
                }
            }
        } catch (Exception e) {
            throw new AppException("Invalid Webhook JSON payload", HttpStatus.BAD_REQUEST);
        }

        return Map.of("received", true);
    }

    public static String calculateHmacSha256(String data, String secret) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKeySpec = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            mac.init(secretKeySpec);
            byte[] hash = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(hash);
        } catch (Exception e) {
            throw new RuntimeException("Failed to calculate HMAC-SHA256 signature", e);
        }
    }
}
